src/viadot/orchestration/prefect/utils.py

```diff
@@ -28,6 +28,394 @@
+        - "X years/months/days ago full date":  e.g., "3 years ago full date", refers to a given date X units ago in dynamic_date_format
+        - "last X years/months/days": e.g., "last 10 months", refers to a data range of the months in 'YMM' format
+        - "Y years from X": e.g., "10 years from 2020", refers to a data range of the year numbers from a specified year
+        - "first X days from X": e.g., "first 10 days of January 2020", returns a data range of days from a given month
+
+        Args:
+            dynamic_date_symbols (List[str], optional): The symbols that mark the start and the end of a dynamic date pattern in a text. Defaults to ["<<", ">>"].
+            dynamic_date_format (str, optional): A date and time format string defining the text representation of date. Defaults to "%Y%m%d".
+            dynamic_date_timezone (str, optional): A string that sets the default timezone used by all datetime functions. Defaults to "Europe/Warsaw".
+        """
+        self.singular_patterns = {
+            "last_day_of_month": r"last\s+day\s+of\s+(\w+)\s+(\d{4})",
+            "x_units_ago_full_date": r"(\d+)\s+(years?|months?|days?)\s+ago\s+full\s+date",
+            "x_years_ago": r"(\d+)\s+years\s+ago",
+        }
+        self.range_patterns = {
+            "last_x_units": r"last\s+(\d+)\s+(years|months|days)",
+            "y_years_from_x": r"(\d+)\s+years\s+from\s+(\d{4})",
+            "first_x_days_from": r"first\s+(\d+)\s+days\s+from\s+(\w+)\s+(\d{4})",
+            "last_x_days_from": r"last\s+(\d+)\s+days\s+from\s+(\w+)\s+(\d{4})",
+        }
+        self.dynamic_date_format = dynamic_date_format
+        self.dynamic_date_timezone = dynamic_date_timezone
+        self.dynamic_date_symbols = dynamic_date_symbols
+        self.replacements = self.create_date_dict()
+
+    def generate_years(
+        self, last_years: int, from_year: str, num_years: str
+    ) -> List[str]:
+        """
+        Generate a list of years either for the last X years or from a start year.
+
+        Args:
+            last_years (int): The number of years from the current year to include.
+            from_year (str): The starting year.
+            num_years (int): The number of years to generate from the starting year.
+
+        Returns:
+            list: A list of years in ascending order.
+        """
+        current_year = pendulum.now().year
+        if last_years:
+            result = [str(current_year - i) for i in range(last_years)][
+                ::-1
+            ]  # Reversed to ascending order
+            return result
+        elif from_year and num_years:
+            start_year = int(from_year)
+            result = [str(start_year - i) for i in range(num_years)][
+                ::-1
+            ]  # Ascending order
+            return result
+        return []
+
+    def generate_months(self, last_months: int) -> List[str]:
```

Comment:
mark this function internal by adding `_`.
```suggestion
    def _generate_months(self, last_months: int) -> List[str]:
```